### ISSUE TITLE
[SUBGRAPH] Rewrite Subgraph Tests as Unit Tests w/ Matchstick OR Huge Refactor of Existing Tests

### ISSUE THREAD

**Issue by 0xdavinchee** [superfluid-finance/protocol-monorepo#1106]
> Note: This is not a priority now, but I just wanted to note this as this has been something that has been bothering me for a bit and triggered by a recent experience with jumping back into the tests.

## Summary
The current Subgraph tests what happens when we actually trigger contract events in a local blockchain with a local subgraph and sees how the Subgraph handles the indexing over time with the same state.

We essentially do this:
- Initialize empty state
- Modify state with known parameters
- Locally modify state in a way we expect the subgraph to modify the state => this is done by essentially duplicating the mapping logic, but in the tests (pretty much no way of getting around duplicating the subgraph mapping logic here)
- Wait for the Subgraph to index the state modification transaction 
- Query the subgraph for the newly indexed 

I am proposing this because I think that this test suite is overkill and the maintenance costs are far too high to justify not trying to do this in a different way which may provide just as much value in a cleaner way. I don't think we have to or should throw these tests away completely, but it might be a worthy investment to try out [matchstick](https://github.com/LimeChain/matchstick) and incrementally add unit tests there to see if makes sense to make a full transition.

As the creator of these tests I have had to go back to work on these tests a few times and it is not easy to follow due to a few reasons:
- how I architected the tests
- the amount of state stored and logic replicated from the subgraph code
- Unrelated, but other people who have worked on it (Marten and Mehul) have also said it's pretty hard to follow

I think we should strongly consider either refactoring this greatly to make it more maintainable by either reducing the scope OR incrementally rewriting some of the tests with a different framework.

**Comment by 0xdavinchee** [comment]
TODOS

- [x] Figure out how to matchstick
- [x] Come up with a plan of how what the unit tests will consist of: test the different mapping functions/creation of entities
  - [x] This should be as simple as understanding what events create what entities and ensuring that this is done as expected
- [ ] Come up with a plan of how what the integration tests will consist of: test that the subgraph is properly indexing and creating an event
  - [ ] Delete the state persistence stuff and just check that all the entities are created when a specific event is emitted and retrievable
- [ ] Figure out how to organize the tests in a way so that it is most maintainable 
- [ ] Write the unit tests
- [ ] Purge a bunch of the integration test code
- [ ] Handle the CI jazz
- [ ] Profit for future me/other devs so they don't hate me as much I hate myself when working on tests here 🥲

**Comment by 0xdavinchee** [comment]
## Matchstick Unit Tests

### Invariant Tests

These tests should be added as a separate describe in each of the test files, we don't need to check all the properties, but only that an entity with the id is created.

- [ ] Whenever a `Token` is created, `TokenStatistic` should be created
- [ ] Whenever `TokenStatistic` is created/updated `TokenStatisticLog` should be created
- [ ] Whenever `Stream` is created/updated/deleted new `StreamPeriod' should be created
- [ ] Whenever `Account` is created `AccountTokenSnapshot` should be created
- [ ] Whenever `AccountTokenSnapshot` is created/updated, `AccountTokenSnapshotLog` should be created
- [ ] Whenever `AccountTokenSnapshot` is updated, we want to update the `*UntilUpdatedAt` properties 


- [x] **host.ts**
- [x] `handleGovernanceReplaced`
    - [x] `GovernanceReplacedEvent`
- [x] `handleAgreementClassRegistered`
    - [x] `AgreementClassRegisteredEvent`
- [x] `handleAgreementClassUpdated`
    - [x] `AgreementClassUpdatedEvent`
- [x] `handleSuperTokenFactoryUpdated`
    - [x] `SuperTokenFactoryUpdatedEvent`
- [x] `handleSuperTokenLogicUpdated`
    - [x] `SuperTokenLogicUpdatedEvent`
- [x] `handleAppRegistered`
    - [x] `AppRegisteredEvent`
- [x] `handleJail`
    - [x] `JailEvent`

- [x] **superfluidGovernance.ts**
- [x] `handleConfigChanged`
    - [x] `ConfigChangedEvent`
- [x] `handleRewardAddressChanged`
    - [x] `RewardAddressChangedEvent`
- [x] `handleCFAv1LiquidationPeriodChanged`
    - [x] `CFAv1LiquidationPeriodChangedEvent`
- [x] `handlePPPConfigurationChanged`
    - [x] `PPPConfigurationChangedEvent`
- [x] `handleTrustedForwarderChanged`
    - [x] `TrustedForwarderChangedEvent`

- [x] **resolver.ts**
~- [ ] `handleRoleAdminChanged`~
    ~- [ ] `RoleAdminChangedEvent`~
~- [ ] `handleRoleGranted`~
    ~- [ ] `RoleGrantedEvent`~
~- [ ] `handleRoleRevoked`~
    ~- [ ] `RoleRevokedEvent`~

> NOTE: The above events come from OpenZeppelin's `AcessControl.sol` contract which `Resolver.sol` inherits and might not be events we actually care about/want to index.

- [x] `handleSet`
    - [x] `SetEvent`
    - [x] `ResolverEntry`
    - [x] `Token`

- [ ] **cfav1.ts** 
- [ ] `handleFlowUpdated`:
    - [x] `FlowUpdatedEvent`
	- [ ] `Stream`
	- [ ] `StreamRevision`
	- [ ] `FlowUpdatedEvent`
	- [ ] `StreamPeriod`
	- [ ] `Account`
	        - Sender
	        - Receiver
	- [ ] `AccountTokenSnapshot`
		- Sender
		- Receiver
	- [ ] `AccountTokenSnapshotLog`
		- Sender
		- Receiver
	- [ ] `TokenStatistic`
	- [ ] `TokenStatisticLog`
	- Tests
		- Mocked create flow event
		- Mocked update flow event
		- Mocked delete flow event
> REMINDER: Pay special attention to the aggregate entities.

- [ ] `handleFlowUpdatedExtension`:
	- [ ] `FlowUpdatedEvent`
    - [ ] `FlowOperator`
	- NOTE: We need to create a `FlowUpdatedEvent` prior to testing this event

- [x] `handleFlowOperatorUpdated`:
    - [x] `FlowOperatorUpdatedEvent`
    - [x] `FlowOperator`

- [ ] **idav1.ts**
- [ ] `handleIndexCreated`
    - [x] `IndexCreatedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For publisher
- [x] `handleIndexDistributionClaimed`
    - [x] `IndexDistributionClaimedEvent`
- [ ] `handleIndexUpdated`
    - [x] `IndexUpdatedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For publisher
- [x] `handleIndexSubscribed`
    - [x] `IndexSubscribedEvent`
- [ ] `handleIndexUnitsUpdated`
    - [x] `IndexUnitsUpdatedEvent`
    - [ ] `Subscription`
    - [ ] `Index`
- [x] `handleIndexUnsubscribed`
    - [x] `IndexUnsubscribedEvent`
- [ ] `handleSubscriptionApproved`
    - [x] `SubscriptionApprovedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For subscriber and publisher (if hasSubWithUnits)
- [ ] `handleSubscriptionDistributionClaimed`
    - [x] `SubscriptionDistributionClaimedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For subscriber and publisher
- [ ] `handleSubscriptionRevoked`
    - [x] `SubscriptionRevokedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For subscriber and publisher
- [ ] `handleSubscriptionUnitsUpdated`
    - [x] `SubscriptionUnitsUpdatedEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For subscriber and publisher

- [ ] **superToken.ts**

- [ ] `handleAgreementLiquidatedBy`:
    - [x] `AgreementLiquidatedByEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For liquidator, bond account and target account
- [ ] `handleAgreementLiquidatedV2`:
    - [x] `AgreementLiquidatedV2Event`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For liquidator, bond account and target account
- [ ] `handleTokenUpgraded`:
    - [x] `TokenUpgradedEvent`
    - [ ] `Account`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For account

- [ ] `handleTokenDowngraded`:
    - [x] `TokenDowngradedEvent`
    - [ ] `Account`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For account
- [ ] `handleTransfer`:
    - [x] `TransferEvent`
    - [ ] `TokenStatistic`
    - [ ] `TokenStatisticLog`
    - [ ] `AccountTokenSnapshot`
    - [ ] `AccountTokenSnapshotLog`
    - NOTE: For from and to
- [x] `handleSent`:
    - [x] `SentEvent`
- [x] `handleBurned`:
    - [x] `BurnedEvent`
    - [x] `TokenStatistic`
    ~- [ ] `TokenStatisticLog`~
- [x] `handleMinted`:
    - [x] `MintedEvent`
    - [x] `TokenStatistic`
    ~- [ ] `TokenStatisticLog`~

- [x] **superTokenFactory.ts**

- [x] `handleSuperTokenCreated`/`handleCustomSuperTokenCreated`:
    - [x] `SuperTokenCreatedEvent`/`CustomSuperTokenCreated`
    - [x] `Token`
        - [x] `Token` for underlying if exists
    - [x] `TokenStatistic`
    - [x] `TokenStatisticLog`

- [x] `handleSuperTokenLogicCreated`:
    - [x] `SuperTokenLogicCreatedEvent`

**Comment by hellwolf** [comment]
is this one outdated?

**Comment by 0xdavinchee** [comment]
kinda outdated, but mainly just not spending any time on writing the tests out

**Comment by hellwolf** [comment]
@vmichalik I suggest we icebox or close this.
